nomad/volumewatcher/volume_watcher.go

```diff
@@ -0,0 +1,382 @@
+	if err != nil {
+		vw.logger.Error("could not query plugins for volume", "error", err)
+		return nil
+	}
+
+	vol, err = vw.state.CSIVolumeDenormalize(ws, vol)
+	if err != nil {
+		vw.logger.Error("could not query allocs for volume", "error", err)
+		return nil
+	}
+	vw.v = vol
+	return vol
+}
+
+// volumeReap collects errors for logging but doesn't return them
+// to the main loop.
+func (vw *volumeWatcher) volumeReap(vol *structs.CSIVolume) {
+	vw.logger.Trace("releasing unused volume claims", "id", vol.ID, "namespace", vol.Namespace)
+	err := vw.volumeReapImpl(vol)
+	if err != nil {
+		vw.logger.Error("error releasing volume claims", "error", err)
+	}
+	if vw.isUnclaimed(vol) {
+		vw.Stop()
+	}
+}
+
+func (vw *volumeWatcher) isUnclaimed(vol *structs.CSIVolume) bool {
+	return len(vol.ReadClaims) == 0 && len(vol.WriteClaims) == 0 && len(vol.PastClaims) == 0
+}
+
+func (vw *volumeWatcher) volumeReapImpl(vol *structs.CSIVolume) error {
+	var result *multierror.Error
+	nodeClaims := map[string]int{} // node IDs -> count
+	jobs := map[string]bool{}      // namespace+jobID -> stopped
+
+	// if a job is purged, the subsequent alloc updates can't
+	// trigger a GC job because there's no job for them to query.
+	// Job.Deregister will send a claim release on all claims
+	// but the allocs will not yet be terminated. save the status
+	// for each job so that we don't requery in this pass
+	checkStopped := func(jobID string) bool {
+		namespace := vw.v.Namespace
+		isStopped, ok := jobs[namespace+jobID]
+		if !ok {
+			ws := memdb.NewWatchSet()
+			job, err := vw.state.JobByID(ws, namespace, jobID)
+			if err != nil {
+				isStopped = true
+			}
+			if job == nil || job.Stopped() {
+				isStopped = true
+			}
+			jobs[namespace+jobID] = isStopped
+		}
+		return isStopped
+	}
+
+	collect := func(allocs map[string]*structs.Allocation,
+		claims map[string]*structs.CSIVolumeClaim) {
+
+		for allocID, alloc := range allocs {
+
+			if alloc == nil {
+				_, exists := vol.PastClaims[allocID]
+				if !exists {
+					vol.PastClaims[allocID] = &structs.CSIVolumeClaim{
+						AllocationID: allocID,
+						State:        structs.CSIVolumeClaimStateReadyToFree,
+					}
+				}
+				continue
+			}
+
+			nodeClaims[alloc.NodeID]++
+
+			if alloc.Terminated() || checkStopped(alloc.JobID) {
+				// don't overwrite the PastClaim if we've seen it before,
+				// so that we can track state between subsequent calls
+				_, exists := vol.PastClaims[allocID]
+				if !exists {
+					claim, ok := claims[allocID]
+					if !ok {
+						claim = &structs.CSIVolumeClaim{
+							AllocationID: allocID,
+							NodeID:       alloc.NodeID,
+						}
+					}
+					claim.State = structs.CSIVolumeClaimStateTaken
+					vol.PastClaims[allocID] = claim
+				}
+			}
+		}
+	}
+
+	collect(vol.ReadAllocs, vol.ReadClaims)
+	collect(vol.WriteAllocs, vol.WriteClaims)
+
+	if len(vol.PastClaims) == 0 {
+		return nil
+	}
+
+	for _, claim := range vol.PastClaims {
```

Comment:
Note for review: from here down we're almost entirely copying the code over from the `nomad/core_sched.go` file that's been removed. Changes mostly include better trace logging and naming/comments.